libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java

```diff
@@ -136,6 +138,25 @@ public boolean isLastLeg(RouteProgress routeProgress) {
     return coordinates;
   }
 
+  /**
+   * Given a {@link RouteProgress}, this method will calculate the remaining waypoint names
+   * along the given route based on route option waypoint names and the progress remaining coordinates.
+   *
+   * @param routeProgress for route waypoint names and remaining coordinates
+   * @return String array of remaining waypoint names
+   * @since 0.19.0
+   */
+  @Nullable
+  public String[] calculateRemainingWaypointNames(RouteProgress routeProgress) {
+    RouteOptions routeOptions = routeProgress.directionsRoute().routeOptions();
+    if (routeOptions == null || TextUtils.isEmpty(routeOptions.waypointNames())) {
+      return null;
```

Comment:
We should consider returning a [Null Object](https://en.wikipedia.org/wiki/Null_Object_pattern#Java) instead or even better try to getting rid of it at all (maybe returning an empty list?). If not, code becomes defensive and full of `null` checks.

👀 [Null References: The Billion Dollar Mistake](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare)